File: apps/android/src/main/kotlin/now/shouldigooutside/widget/WidgetColors.kt
package now.shouldigooutside.widget

import androidx.compose.ui.graphics.Color
import androidx.glance.unit.ColorProvider
import now.shouldigooutside.core.model.score.ScoreResult
import now.shouldigooutside.core.ui.Colors
import now.shouldigooutside.core.ui.DarkColors
import now.shouldigooutside.core.ui.LightColors
import androidx.glance.color.ColorProvider as DayNightColorProvider

internal fun widgetColors(isDark: Boolean): Colors = if (isDark) DarkColors else LightColors

internal fun Colors.scoreColor(result: ScoreResult): Color =
    when (result) {
        ScoreResult.Yes -> success
        ScoreResult.Maybe -> primary
        ScoreResult.No -> error
    }

internal fun Color.toProvider(): ColorProvider = DayNightColorProvider(this, this)

internal val BlackProvider: ColorProvider = Color.Black.toProvider()
internal val WhiteProvider: ColorProvider = Color.White.toProvider()
internal val BlackLocationProvider: ColorProvider =
    Color.Black.copy(alpha = WidgetDimens.LOCATION_ALPHA).toProvider()
internal val BlackActivityProvider: ColorProvider =
    Color.Black.copy(alpha = WidgetDimens.ACTIVITY_ALPHA).toProvider()
internal val BlackStaleProvider: ColorProvider =
    Color.Black.copy(alpha = WidgetDimens.STALE_ALPHA).toProvider()

File: apps/ios/iosApp/Shared/WidgetColors.swift
import SwiftUI

/// Widget colors mirrored from core/ui Color.kt LightColors/DarkColors.
/// Update when the app theme colors change.
struct WidgetThemeColors {
    let success: Color
    let primary: Color
    let error: Color
    let surface: Color
    let text: Color
    let textSecondary: Color
    let onSuccess: Color

    func scoreColor(for result: ScoreResult) -> Color {
        switch result {
        case .Yes: return success
        case .Maybe: return primary
        case .No: return error
        }
    }
}

private let lightColors = WidgetThemeColors(
    success: Color(hex: 0xFF90EE90),
    primary: Color(hex: 0xFFFFDB02),
    error: Color(hex: 0xFFFF6B6B),
    surface: Color(hex: 0xFFFFF5EA),
    text: .black,
    textSecondary: Color(hex: 0xFF333333),
    onSuccess: .black
)

private let darkColors = WidgetThemeColors(
    success: Color(hex: 0xFF7FFFD4),
    primary: Color(hex: 0xFFFCCF03),
    error: Color(hex: 0xFFFF8C91),
    surface: Color(hex: 0xFF2A2236),
    text: Color(hex: 0xFFF8F8F8),
    textSecondary: Color(hex: 0xFFAAAAAA),
    onSuccess: .black
)

func widgetColors(scheme: ColorScheme) -> WidgetThemeColors {
    scheme == .dark ? darkColors : lightColors
}

extension Color {
    /// Text laid over a score color background. Black at varying alphas mirrors
    /// the Android widget's Black* color providers.
    static let widgetOnScore = Color.black
    static let widgetOnScoreLocation = Color.black.opacity(0.8)
    static let widgetOnScoreTitle = Color.black.opacity(0.7)
    static let widgetOnScoreStale = Color.black.opacity(0.6)
}

private extension Color {
    init(hex: UInt64) {
        let a = Double((hex >> 24) & 0xFF) / 255.0
        let r = Double((hex >> 16) & 0xFF) / 255.0
        let g = Double((hex >> 8) & 0xFF) / 255.0
        let b = Double(hex & 0xFF) / 255.0
        self.init(red: r, green: g, blue: b, opacity: a)
    }
}
